src/ADC.c

```diff
@@ -1,69 +1,380 @@
+
+#define GPIO_AFRL_AFRL1_0 (0x00000010)
+#define GPIO_AFRL_AFRL1_1 (0x00000020)
+#define GPIO_AFRL_AFRL1_2 (0x00000040)
+#define GPIO_AFRL_AFRL1_3 (0x00000080)
+
+/* Clock rate of the timer */
+#define TIMER_CLOCK_RATE (67600000)
+
+/* 9 cycles at 67.6MHz gives 7.51MHz which is a bit less than 8MHz */
+#define EXT_CLOCK_PERIOD (9)
+
+/* CONVST PWM with computed period and a on time of 37 cycles */
+#define CONVERSION_TIME (30)
+#define ACQUISITION_TIME (1)
+#define CONVST_PERIOD (((CONVERSION_TIME + ACQUISITION_TIME) * EXT_CLOCK_PERIOD) - 1)
+#define CONVST_DUTY (CONVERSION_TIME * EXT_CLOCK_PERIOD)
+#define CONVST_PSC (0)
+
+/* RD PWM with 50% duty, 3.33 MHz frequency */
+#define RD_PERIOD (0xFFFF)
+#define RD_DUTY (EXT_CLOCK_PERIOD + (EXT_CLOCK_PERIOD/2))
+#define RD_PSC (0)
+
+#define ADC_MAX_READINGS (4096)
+#define ADC_CHANNEL_COUNT (4)
+#define ADC_BUFFER_SIZE (ADC_MAX_READINGS * ADC_CHANNEL_COUNT)
+
+uint16_t ADC_Buffer[ADC_BUFFER_SIZE + 1];
+static uint16_t buffer_size = ADC_BUFFER_SIZE + 1;
+static volatile uint8_t ADC_mutex = 1;
+
+static void configure_ADC_GPIO() {
+	/* Configure CS (Chip Select) Pin as output pin PB0 and set it high */
+	RCC->AHB1ENR |= RCC_AHB1ENR_GPIOBEN; /* Enable clock for B pins */
+	GPIOB->MODER |= GPIO_MODER_MODER0_0; /* Output mode */
+	GPIOB->BSRRL = GPIO_BSRR_BS_0; /* Set pin to 1 */
+
+	/* Configure RD (Read) Pin as output pin PB1 and set it high */
+	GPIOB->MODER |= GPIO_MODER_MODER1_0;
+	GPIOB->BSRRL = GPIO_BSRR_BS_1;
+
+	/* Configure WR (Write) Pin as output pin PB2 and set it high */
+	GPIOB->MODER |= GPIO_MODER_MODER2_0; /* Output mode */
+	GPIOB->BSRRL = GPIO_BSRR_BS_2;
+
+	/* CONVST is on PB4 and is configured in configure_CONVST_PWM function */
+
+	/* EOLC is on PB7 and is an input pin by default */
+
+	/* Configure CLK_CNTRL pin as output PB10 and set it low for now */
+	GPIOB->MODER |= GPIO_MODER_MODER10_0; /* Output mode */
+	GPIOB->BSRRH = GPIO_BSRR_BR_10 >> 16;
+	/* Should be high for external clock, do this after ADC in configured */
+
+	/* Enable RCC for GPIOA for the external clock to the ADC on pin PA8 */
+	RCC->AHB1ENR |= RCC_AHB1ENR_GPIOAEN; /* Enable clock for A pins */
+
+	/* EOC is on PA9 and is an input pin by default */
+
+	/* Configure the ADC digital input pins (all inputs by default) */
+	RCC->AHB1ENR |= RCC_AHB1ENR_GPIOCEN; /* Enable clock for C pins */
+	GPIOC->PUPDR |= GPIO_PUPDR_PUPDR14_1 | GPIO_PUPDR_PUPDR15_1; /* Pull down pins PD14 and PD15 */
+}
+
+/* Configure the ADC to use 4 analog channels */
+static void configure_ADC_IC() {
+	/* Temporarily make pins PC0 - PC7 output pins to configure ADC */
+	GPIOC->MODER |=
+		GPIO_MODER_MODER0_0 | GPIO_MODER_MODER1_0 | GPIO_MODER_MODER2_0 | GPIO_MODER_MODER3_0 |
+		GPIO_MODER_MODER4_0 | GPIO_MODER_MODER5_0 | GPIO_MODER_MODER6_0 | GPIO_MODER_MODER7_0;
+
+	/* Drive WR low and CS low */
+	GPIOB->BSRRH = GPIO_BSRR_BR_2 >> 16;
+	GPIOB->BSRRH = GPIO_BSRR_BR_0 >> 16;
+
+	/* Write configuration to GPIO C */
+	GPIOC->ODR = 0x000F;
```

Comment:
I'm guessing this is to set pins PC0 ro PC7 high? I would prefer if you used the standard defines and OR'd them together.

Reply:
Sure thing! Must have just rushed this :P

Reply:
Also, PC0 - PC3 (all the more reason the defines are probably better)